lib/elixir/src/elixir_clauses.erl

```diff
@@ -40,8 +40,18 @@ guard({'when', Meta, [Left, Right]}, E) ->
   {ELeft, EL}  = guard(Left, E),
   {ERight, ER} = guard(Right, EL),
   {{'when', Meta, [ELeft, ERight]}, ER};
-guard(Other, E) ->
-  elixir_expand:expand(Other, E).
+guard(Guard, E) ->
+  warn_zero_length_guard(Guard, E),
```

Comment:
We should check for the warning on the expanded guard. That's because the user can unimport Kernel's length and bring their own length which may not have the issue below!